File: backend/kotlin/backend-kotlin-spring/src/main/kotlin/dev/reference/app/controller/AuthController.kt
package dev.reference.app.controller

import dev.reference.app.dto.AuthResponse
import dev.reference.app.dto.LoginRequest
import dev.reference.app.dto.RegisterRequest
import dev.reference.app.dto.UserProfileResponse
import dev.reference.app.exception.AuthException
import dev.reference.app.service.AuthService
import jakarta.validation.Valid
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.security.core.annotation.AuthenticationPrincipal
import org.springframework.web.bind.annotation.ExceptionHandler
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.RequestBody
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RestController
import org.springframework.web.bind.annotation.RestControllerAdvice

@RestController
@RequestMapping("/api/auth")
class AuthController(
    private val authService: AuthService,
) {
    @PostMapping("/register")
    fun register(@Valid @RequestBody request: RegisterRequest): ResponseEntity<AuthResponse> =
        ResponseEntity.status(HttpStatus.CREATED).body(authService.register(request))

    @PostMapping("/login")
    fun login(@Valid @RequestBody request: LoginRequest): AuthResponse =
        authService.login(request)

    @PostMapping("/logout")
    fun logout(): ResponseEntity<Void> = ResponseEntity.noContent().build()

    @GetMapping("/me")
    fun me(@AuthenticationPrincipal username: String): UserProfileResponse =
        authService.profile(username)
}

@RestControllerAdvice
class AuthExceptionHandler {
    @ExceptionHandler(AuthException::class)
    fun handleAuthException(ex: AuthException): ResponseEntity<Map<String, String>> =
        ResponseEntity.status(ex.status).body(mapOf("message" to requireNotNull(ex.message)))
}

File: backend/javascript/backend-javascript-nest/src/auth.controller.js
'use strict';

const {
  Controller,
  Post,
  Get,
  Body,
  Req,
  HttpCode,
  UseGuards,
} = require('@nestjs/common');

const { AuthService } = require('./auth.service');
const { AuthGuard } = require('./auth.guard');
const { injectConstructor, decorateParams } = require('./di');

@Controller('api/auth')
class AuthController {
  constructor(auth) {
    this.auth = auth;
  }

  @Post('register')
  @HttpCode(201)
  async register(body) {
    return this.auth.register(body);
  }

  @Post('login')
  @HttpCode(200)
  async login(body) {
    return this.auth.login(body);
  }

  @Post('logout')
  @HttpCode(204)
  logout() {
    return undefined;
  }

  @Get('me')
  @UseGuards(AuthGuard)
  me(request) {
    return { username: request.user.username };
  }
}

injectConstructor(AuthController, AuthService);
decorateParams(AuthController, 'register', Body());
decorateParams(AuthController, 'login', Body());
decorateParams(AuthController, 'me', Req());

module.exports = { AuthController };
